Database Streaming

ricdezen opened the pull request:
**fix (`postgresql_adapter.py`):**
**~** Was renamed from `posgresql_adapter`, both the module and the class got fixed. The reference in `upload.py` got fixed too.

**feat (`postgresql_adapter.py`):**
**+** Added `stream(query)` method to return a `PostgreSQLStream` object

**feat (`database_stream.py`):**
**+** **DatabaseStream**, generic Iterable class for database streams: `__next__()`, `__iter__()`, `has_next()`
**+** **PostgreSQLStream**: postgresql implementation:
**+** Requires a named cursor and the query to execute.
**+** Starts the query on the first `__iter__()` call and returns itself
**+** Each `__next__()` call calls `cursor.fetchone()`, when `StopIteration` is raised, closes the cursor and raises it again
**+** `has_next()` returns the opposite of `cursor.closed`
**+** `close()`, allows closing the cursor.

Commit:
fix (run_all_tests.py) : removed stdout capture

Commit:
fix(postgresql_adapter.py):
~ fixed module and class name: from posgresql... to postgresql...

fix(upload.py):
~ fixed reference to renamed PostgreSQLAdapter class and its module

Commit:
feat (postgresql_adapter.py):
+ added stream(query) method returning a DatabaseStream object, specifically of the PostgreSQLStream subclass.

feat (database_stream.py):
+ DatabaseStream: class representing a generic Iterable stream of data
+ PostgreSQLStream: class for a streaming of a postgresql db, requires a server-side/named cursor and a query
+ basic iterable implementation: __next__, __iter__, has_next
+ The stream can be used a single time, after which the cursor is closed and it becomes unusable

Commit:
feat (profile_batch.py): just a utility file to help profiling the database streaming based on tha batch size

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -0,0 +1,80 @@
+
+class DatabaseStream():
+
+    def __iter__(self):
```
Is this class an `Iterator` or an `Iterable`? You shouldn't mix the two things by using both `__iter__` and `__next__`

CremaLuca commented on otri/database/postgresql_adapter.py:
```diff
@@ -1,11 +1,13 @@
+from typing import Iterable
 
 
-class PosgreSQLAdapter(DatabaseAdapter):
```
Nice, I'm a freaking dyslexic

Commit:
fix (profile_batch.py): removed useless .close() call
fix (postgresql_adapter.py): fixed call to PostgreSQLStream
feat(database_stream.py): divided Stream from Iterator

Commit:
fix : codacy style fixes
fix (database_stream.py):
fixed missing return statement in PostgreSQLStream.__next__()

Commit:
fix (profile_batch.py) : minor addition to the output log

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -0,0 +1,117 @@
+            return next(self.__cursor)
+        except StopIteration:
+            self.__cursor.close()
+            raise StopIteration
```
You handled the cursor thing excellently, couldn't have done better

CremaLuca commented on otri/database/postgresql_adapter.py:
```diff
@@ -72,9 +73,25 @@ def read(self, query: DatabaseQuery):
+                Executes the query on the given query.category and the given query.filters
+            batch_size : int
+                The number of rows the database should load before making them available.
+                The iterable still always yealds a single item.
```
```suggestion
                The iterable still always yields a single item.
```

CremaLuca commented on run_all_tests.py:
```diff
@@ -1,3 +1,3 @@
 
-os.system('cmd /c "python -m pytest"')
\ No newline at end of file
+os.system('cmd /c "python -m pytest -s"')
```
What does this -s do?

ricdezen replied:
Removes the capturing of the print() statements, to allow logging

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -0,0 +1,117 @@
+            self.__cursor.close()
+            raise StopIteration
+
+    def has_next(self):
```
```suggestion
    def has_next(self) -> bool:
```

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -0,0 +1,117 @@
+        self.__query = query
+        self.__batch_size = batch_size
+
+    def __iter__(self):
```
```suggestion
    def __iter__(self) -> DatabaseIterator:
```

CremaLuca reviewed (approved):
Very noice, seems almost ready to be merged

ricdezen committed:
Update otri/database/postgresql_adapter.py

Co-authored-by: Luca Crema <lc.crema@hotmail.com>

ricdezen committed:
fix (database_stream.py) : missing return type hint

Co-authored-by: Luca Crema <lc.crema@hotmail.com>

CremaLuca commented:
Now it's-a perfect.
![Perfect](https://media3.giphy.com/media/3o8doT9BL7dgtolp7O/giphy.gif?cid=ecf05e4765c7129d2d6f9766e56a28b9cdb46754fe39f450&rid=giphy.gif)
